README.md
# Game Levels Scrolling Map

A package for making game levels map like candy crush or similar games using flutter with ability to be horizontal or vertical

## Getting Started

* This package enables you to build a Game levels map like candy crush with the following features :
  - Option to make it horizontal map or vertical
  - Option to reverse the scrolling start direction
  - Option to add the x,y points positions
  - Option to extract the x,y points positions from asset SVG file or online SVG file

## Platform Support

| Android ✔️ | IOS ✔️ | MacOS ✔️ ️| Web ✔️ | Linux ✔️ | Windows ✔ ️|

## Usage
<br>

**Step 1** Import `package:game_levels_scrolling_map/game_levels_scrolling_map.dart` which contain the main widget of the map `GameLevelsScrollingMap`
then Import `package:game_levels_scrolling_map/model/point_model.dart` which contain the model of points<br>

**Step 2** Start adding your points widgets by creating a new List of points and adding widgets to it using the `PointModel` class<br>

```dart
List<PointModel> points = [];
      
for(int i = 0; i<50 ; i++){
  PointModel point = PointModel(100,Container(width: 40, height: 40, color: Colors.red, child: Text("$i")));
  
  /* To make the map scroll to a specific point just make its parameter 'isCurrent' = true like the following which will make the map scroll to it once created*/
  if(i == 20) point.isCurrent = true;
  
  points.add(point);
}
```

**Step 3** Use `GameLevelScrollingMap.scrollable` widget to build the map<br>

**- To make vertical map:**<br>
<img src="https://raw.githubusercontent.com/MohamedSayed9392/Game-Levels-Scrolling-Map/master/screenshots/screenshot_map_vertical.png" alt="Vertical Map Example" width="400"/><br>

```dart
Widget build(BuildContext context) {
  return Scaffold(
    body: Container(
      child: GameLevelsScrollingMap.scrollable(
              imageUrl: "assets/drawable/map_vertical.png",
              direction: Axis.vertical,
              reverseScrolling: true,
              svgUrl: 'assets/svg/map_vertical.svg',
              points: points,)
          ),
        );
}
```
**- To make horizontal map:**<br>
<img src="https://raw.githubusercontent.com/MohamedSayed9392/Game-Levels-Scrolling-Map/master/screenshots/screenshot_map_horizantal.png" alt="Horizontal Map Example" width="500"/><br><br>

```dart
Widget build(BuildContext context) {
  return Scaffold(
    body: Container(
      child: GameLevelsScrollingMap.scrollable(
            imageUrl: "assets/drawable/map_horizontal.png",
            direction: Axis.horizontal,
            svgUrl: 'assets/svg/map_horizontal.svg',
            points: points,)
          ),
        );
}
```





CHANGELOG.md
## 0.0.7

* Add `imageHeight` and `imageWidth` attributes if you know exactly the real height and real width of the background map image
* Add `backgroundImageWidget` attributes if you need to pass the background as widget (Ex. if you need to use your own image widget instead of Image.asset)
* Fix some responsiveness issues on small phones

## 0.0.6

* Add compatibility for `Linux` and `MacOS`
* Add 'isCurrent' attribute to PointModel which make the scorlling map scroll to a specific point like the current level in games

## 0.0.5

* Add compatibility for `Web` and `Windows`

## 0.0.4

* Add 2 optional parameters `pointsPositionDeltaX` and `pointsPositionDeltaY` to manually adjust the points in case of the points is shifted in `x` or `y` axis

## 0.0.3

* Game levels map like candy crush with the following features :
  - Option to make it horizontal map or vertical
  - Option to reverse the scrolling start direction
  - Option to add the x,y points positions
  - Option to extract the x,y points positions from asset SVG file or online SVG file

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'examples/map_vertical_example.dart';
import 'examples/map_horizontal_example.dart';

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await SystemChrome.setPreferredOrientations([DeviceOrientation.landscapeLeft]);
  await SystemChrome.setEnabledSystemUIMode(SystemUiMode.manual, overlays: []);

  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {


  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MapHorizontalExample(),
      //home: MapVerticalExample(),
    );
  }
}


lib/helper/utils.dart

import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:path_provider/path_provider.dart';
import 'package:xml/xml.dart';

class Utils {

  static Future<String> readFile(String filePath, {isTest = false}) async {
    return isTest || !filePath.contains("assets")
        ? File(filePath).readAsString()
        : rootBundle.loadString(filePath);
  }

  // How I read SVG files
  static Future<XmlDocument> readSvg(String filePath, {isTest = false}) async {
    String path = filePath;
    if(filePath.contains("http")){
      path = await downloadFile(filePath);
    }
    var content = await readFile(path, isTest: isTest);
    var xmlRoot = XmlDocument.parse(content);
    return xmlRoot;
  }

  static Future<String> downloadFile(String url) async{
    final request = await HttpClient().getUrl(Uri.parse(url));

    final response = await request.close();
    Directory tempDir = await getTemporaryDirectory();
    String path = "${tempDir.path}/temp_file";
    await response.pipe(File(path).openWrite());
    return path;
  }

// How I filter out specific elements
  static Iterable<XmlElement> getXmlWithClass(XmlDocument root, String classId) {
    return root.descendants
        .whereType<XmlElement>()
        .where((p) => p.getAttribute('class') == classId)
        .toList();
  }

  static List<Offset> pathToCoords(Path path, [double pixelGap = 2]) {
    var metrics = path.computeMetrics();
    if (metrics.isEmpty) return [];

    var coords = <Offset>[];
    var metric = path.computeMetrics().first;

    double position = 0;
    while (position < metric.length) {
      var tangent = metric.getTangentForOffset(position);
      coords.add(tangent!.position);
      position += pixelGap;
    }

    return coords;
  }
}


lib/widgets/loading_progress.dart
import 'package:flutter/material.dart';

class LoadingProgress extends StatelessWidget {
  const LoadingProgress({ Key? key }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: () {
        print("progress : onTap");
      },
      child: Container(
          color: Colors.black38,
          width:  double.infinity,
          height: 150,
          child: Center(
              child: Card(
                  elevation: 10,
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(15.0),
                  ),
                  child: const Padding(
                    padding: EdgeInsets.all(20.0),
                    child: SizedBox(
                      child: CircularProgressIndicator(
                        strokeWidth: 6,
                        valueColor: AlwaysStoppedAnimation<Color>(
                            Colors.lightBlueAccent),
                      ),
                      height: 70,
                      width: 70,
                    ),
                  )))),
    );
  }
}

lib/game_levels_scrolling_map.dart
library game_levels_scrolling_map;

import 'dart:async';
import 'dart:ui' as ui show Image;

import 'package:flutter/material.dart';
import 'package:xml/xml.dart';

import '../helper/utils.dart';
import 'config/q.dart';
import 'model/point_model.dart';
import 'widgets/loading_progress.dart';

class GameLevelsScrollingMap extends StatefulWidget {
  double? height;
  double? width;

  double? imageHeight;
  double? imageWidth;

  double? currentPointDeltaY;
  String imageUrl;
  String? svgUrl = "";

  List<double>? x_values = [];
  List<double>? y_values = [];

  List<PointModel>? points = [];

  double? pointsPositionDeltaX = 0;
  double? pointsPositionDeltaY = 0;

  bool isScrollable = false;
  Axis? direction = Axis.horizontal;
  bool? reverseScrolling = false;

  Widget? backgroundImageWidget;

  GameLevelsScrollingMap({
    this.imageUrl = "",
    @required this.width,
    this.imageWidth,
    this.imageHeight,
    this.svgUrl,
    this.points,
    this.x_values,
    this.y_values,
    this.pointsPositionDeltaX,
    this.pointsPositionDeltaY,
    this.currentPointDeltaY,
    this.backgroundImageWidget,
    Key? key,
  }) : super(key: key);

  GameLevelsScrollingMap.scrollable( {this.imageUrl = "",
    this.width,
    this.height,
    this.imageWidth,
    this.imageHeight,
    this.direction,
    this.currentPointDeltaY,
    this.svgUrl,
    this.points,
    this.x_values,
    this.y_values,
    this.pointsPositionDeltaX,
    this.pointsPositionDeltaY,
    this.reverseScrolling,
    this.backgroundImageWidget,
    Key? key,
  }) : super(key: key) {
    isScrollable = true;
  }

  @override
  _GameLevelsScrollingMapState createState() => _GameLevelsScrollingMapState();
}

class _GameLevelsScrollingMapState extends State<GameLevelsScrollingMap> {
  List<double>? newX_values = [];

  @override
  void initState() {
    super.initState();

    WidgetsBinding.instance!.addPostFrameCallback((_) async {
      initDeviceDimensions();
      initDefaults();

      if (widget.svgUrl!.isNotEmpty) {
        await _getPathFromSVG();
      }
      await _loadImage(path : widget.imageUrl);
      if (widget.isScrollable) {
        int currentIndex =
        widget.points!.indexWhere((point) => point.isCurrent!);
        if (currentIndex != -1 && newX_values!.isNotEmpty) {
          _scrollController.animateTo(newX_values![currentIndex],
              duration: const Duration(milliseconds: 1000),
              curve: Curves.easeIn);
        }
      }
    });
  }

  void initDeviceDimensions() {
    Q.deviceWidth = MediaQuery
        .of(context)
        .size
        .width;
    Q.deviceHeight = MediaQuery
        .of(context)
        .size
        .height;

    print("${Q.TAG} Device Dimensions : ${Q.deviceWidth} x ${Q.deviceHeight}");
  }

  void initDefaults() {
    widget.svgUrl ??= "";
    widget.x_values ??= [];
    widget.y_values ??= [];
    widget.points ??= [];
    widget.pointsPositionDeltaX ??= 0;
    widget.pointsPositionDeltaY ??= 0;
    widget.direction ??= Axis.horizontal;
    widget.reverseScrolling ??= false;
    widget.height ??= Q.deviceHeight;
    if (widget.direction == Axis.vertical) {
      widget.width ??= Q.deviceWidth;
    } else if (widget.direction == Axis.horizontal) {
      widget.width ??= 0;
    }
    if (widget.width == double.infinity) {
      widget.width = Q.deviceWidth;
    }

    print("widget.height : ${widget.height}");
  }

  final _scrollController = ScrollController();
  final _key = GlobalKey();
  List<Widget> widgets = [];

  @override
  Widget build(BuildContext context) {
    return Container(
        alignment: Alignment.center,
        key: widget.key ?? _key,
        color: Colors.red,
        width: widget.width != 0 ? widget.width : maxWidth,
        height: maxHeight,
        child: widget.isScrollable
            ? SingleChildScrollView(
          controller: _scrollController,
          scrollDirection: widget.direction ?? Axis.horizontal,
          reverse: widget.reverseScrolling ?? false,
          child: aspectRatioWidget(),
        )
            : aspectRatioWidget());
  }

  Widget aspectRatioWidget() {
    return AspectRatio(
      aspectRatio: aspectRatio,
      child: LayoutBuilder(
        builder: (BuildContext context, BoxConstraints constraints) {
          return imageWidth != 0
              ? Stack(
              textDirection: TextDirection.ltr,
              alignment: AlignmentDirectional.topStart,
              children: widgets)
              : const LoadingProgress();
        },
      ),
    );
  }

  bool isImageLoaded = false;

  double aspectRatio = 1;
  double imageWidth = 0;
  double imageHeight = 0;
  double maxWidth = 0;
  double maxHeight = 0;

  Completer<ui.Image> completer = Completer<ui.Image>();
  ImageStream? stream;
  ImageStreamListener? listener;
  Future<ui.Image> _loadImage({String path = ""}) async {

    if(path.isEmpty){
      imageStreamListenerCallBack();
    }else {
      if (path.contains("assets")) {
        stream = AssetImage(path).resolve(ImageConfiguration.empty);
      } else {
        stream = NetworkImage(path).resolve(ImageConfiguration.empty);
      }

      listener = ImageStreamListener((ImageInfo frame, bool synchronousCall) {
        imageStreamListenerCallBack(frame:frame,synchronousCall:synchronousCall);
      });

      stream?.addListener(listener!);
    }
    return completer.future;
  }

  ui.Image? image;
  void imageStreamListenerCallBack({ImageInfo? frame, bool? synchronousCall}){
    if(frame != null) {
      image = frame?.image;
    }
    imageWidth = widget.imageWidth ?? image!.width.toDouble();
    imageHeight = widget.imageHeight ?? image!.height.toDouble();

    print("${Q.TAG} image path : ${widget.imageUrl}");
    print("${Q.TAG} image dimensions : $imageWidth x $imageHeight");

    aspectRatio = imageWidth / imageHeight;

    if (widget.isScrollable) {
      if (widget.direction == Axis.horizontal) {
        maxHeight = widget.height!;
        maxWidth = maxHeight * aspectRatio;
      } else if (widget.direction == Axis.vertical) {
        maxWidth = widget.width!;
        maxHeight = maxWidth / aspectRatio;
      }
    } else {
      maxWidth = widget.width!;
      maxHeight = maxWidth / aspectRatio;
    }

    print(
        "${Q
            .TAG} image all dimensions : $imageWidth : $imageHeight : $aspectRatio");
    print(
        "${Q
            .TAG} image new dimensions : $maxWidth : $maxHeight : $aspectRatio");

    widgets.add(backgroundImage());
    drawPoints();

    if(image!= null) {
      completer.complete(image);
      stream?.removeListener(listener!);
    }

    setState(() => {});
  }

  Widget backgroundImage() {
    return Container(
      height: maxHeight,
      width: maxWidth,
      child: imageWidth != 0
          ? widget.backgroundImageWidget ?? (widget.imageUrl.contains("assets")
          ? Image.asset(widget.imageUrl, fit: BoxFit.fill,filterQuality: FilterQuality.high)
          : Image.network(widget.imageUrl, fit: BoxFit.fill,filterQuality: FilterQuality.high))
          : const LoadingProgress(),
    );
  }

  void drawPoints() {
    double halfScreenSize = (MediaQuery
        .of(context)
        .size
        .width) / 2;
    print("${Q.TAG} maxWidth / imageWidth : ${maxWidth / imageWidth}");
    print("${Q.TAG} maxHeight / imageHeight : ${maxHeight / imageHeight}");

    for (int i = 0; i < widget.points!.length; i++) {
      //widget.points!.add(PointModel(100, testWidget(i)));
      if (widget.x_values!.length > i) {
        var x = (widget.x_values![i] * maxWidth / imageWidth) +
            widget.pointsPositionDeltaX!;

        x = x - (widget.points![i].width! / 2);
        newX_values!.add((x - halfScreenSize).abs());

        var y = ((widget.y_values![i] * maxHeight / imageHeight) +
            widget.pointsPositionDeltaY!);
        if (widget.points![i].isCurrent! && widget.currentPointDeltaY != null) {
          y = y - widget.currentPointDeltaY!;
        }
        y = y - (widget.points![i].width! / 2);

        print(
            "${Q.TAG} old x,y : ${widget.x_values![i]},${widget
                .y_values![i]} ## new x,y : $x,$y");
        widgets.add(pointWidget(x, y, child: widget.points![i].child));
      } else {
        break;
      }
    }
  }

  Widget pointWidget(double x, double y, {Widget? child}) {
    return Positioned(child: child ?? Container(), left: x, top: y);
  }

  String? _pathSVG;

  Future _getPathFromSVG() async {
    await getPointsPathFromXML().then((value) {
      _pathSVG = value.replaceAll(",", " ");
      print("pathSVG : $_pathSVG");
      List<String> arrayOfPoints = _pathSVG!.split(" ");
      for (int i = 0; i < arrayOfPoints.length; i++) {
        if (i % 2 == 0) {
          widget.x_values!.add(double.parse(arrayOfPoints[i]));
        } else {
          widget.y_values!.add(double.parse(arrayOfPoints[i]));
        }
      }
    });
  }

  Future<String> getPointsPathFromXML() async {
    String path = "";
    XmlDocument x = await Utils.readSvg(widget.svgUrl!);
    Utils.getXmlWithClass(x, "st0").forEach((element) {
      path = element.getAttribute("points")!;
    });
    return path;
  }
}


example/lib/examples/map_vertical_example.dart
import 'package:flutter/material.dart';
import 'package:game_levels_scrolling_map/game_levels_scrolling_map.dart';
import 'package:game_levels_scrolling_map/model/point_model.dart';

class MapVerticalExample extends StatefulWidget {
  const MapVerticalExample({Key? key}) : super(key: key);

  @override
  State<MapVerticalExample> createState() => _MapVerticalExampleState();
}

class _MapVerticalExampleState extends State<MapVerticalExample> {


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
          child: GameLevelsScrollingMap.scrollable(
            imageUrl:"assets/drawable/map_vertical.png",
            direction: Axis.vertical,
            reverseScrolling: true,
            pointsPositionDeltaX: 25,
            pointsPositionDeltaY: 25,
            svgUrl: 'assets/svg/map_vertical.svg',
            points: points,)
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  @override
  void initState() {
    fillTestData();
  }

  List<PointModel> points = [];

  void fillTestData() {
    for(int i = 0; i<100 ; i++){
      points.add(PointModel(100,testWidget(i)));
    }
  }


  Widget testWidget(int order) {
    return InkWell(
      child: Stack(
        alignment: Alignment.center,
        children: [
          Image.asset(
            "assets/drawable/map_vertical_point.png",
            fit: BoxFit.fitWidth,
            width: 50,
          ),
          Text("$order",
              style: const TextStyle(color: Colors.black,
                  fontSize: 15))
        ],
      ),
      onTap: () {
        showDialog(
          context: context,
          builder: (BuildContext context) {
            return AlertDialog(
              content: Text("Point $order"),
              actions: <Widget>[
                ElevatedButton(
                  child: const Text("OK"),
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                ),
              ],
            );
          },
        );
      },
    );
  }
}

example/lib/examples/map_horizontal_example.dart
import 'package:flutter/material.dart';
import 'package:game_levels_scrolling_map/game_levels_scrolling_map.dart';
import 'package:game_levels_scrolling_map/model/point_model.dart';

class MapHorizontalExample extends StatefulWidget {
  const MapHorizontalExample({Key? key}) : super(key: key);

  @override
  State<MapHorizontalExample> createState() => _MapHorizontalExampleState();
}

class _MapHorizontalExampleState extends State<MapHorizontalExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
          child: GameLevelsScrollingMap.scrollable(
        imageUrl: "assets/drawable/map_horizontal.png",
        svgUrl: "assets/svg/map_horizontal.svg",
        points: points,
      )), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  @override
  void initState() {
    fillTestData();
  }

  List<PointModel> points = [];

  void fillTestData() {
    for (int i = 0; i < 50; i++) {
      points.add(PointModel(100, testWidget(i)));
    }
  }

  Widget testWidget(int order) {
    return InkWell(
      hoverColor: Colors.blue,
      child: Stack(
        alignment: Alignment.center,
        children: [
          Image.asset(
            "assets/drawable/map_horizontal_point.png",
            fit: BoxFit.fitWidth,
            width: 100,
          ),
          Text("$order",
              style: const TextStyle(color: Colors.black, fontSize: 40))
        ],
      ),
      onTap: () {
        showDialog(
          context: context,
          builder: (BuildContext context) {
            return AlertDialog(
              content: Text("Point $order"),
              actions: <Widget>[
                ElevatedButton(
                  child: const Text("OK"),
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                ),
              ],
            );
          },
        );
      },
    );
  }
}
